### DISCUSSION TITLE
Installing in termux

### DISCUSSION THREAD

**Discussion by sandorex** [jcorporation/myMPD#612] [Q&A] [answered]
Hello, i would like to make my old phone a mpd server but i need help how to make myMPD work in termux, any help would be appreciated

Maybe just building for aarch64 and copying it to the phone would work?

**Comment by sandorex** [comment]
Currently the way i could get it to work is by installing it in Fedora 34 chroot using PRoot but the still can't quite get the ports to work properly, will have to fiddle with that..

**Comment by jcorporation** [comment]
I have no experience with termux, but there was an issue in the past: https://github.com/jcorporation/myMPD/issues/552

Eventually this user can help you? @piyushgarg

**Reply by sandorex** [reply]
Thanks for the info, ill try asking them

**Reply by piyushgarg** [reply]
@sandorex 
Once you are able to build it (unfortunately not able to build with lua but should not cause any issue with myMPD),
clone this repo

```
cd myMPD
export MYMPD_INSTALL_PREFIX=$PREFIX
./build.sh release

```

You can copy it from one device to another device running termux (arch should be same and it will work just fine)
then you should be able to start it like this (change env vars accordingly, after first run the env vars are copied to myMPD config files). 

```
export MYMPD_HTTP_PORT=8080
export MYMPD_LOGLEVEL=6
export MPD_HOST=localhost
export MPD_PORT=6600
export MYMPD_SSL=true
export MYMPD_SSL_PORT=443

<path>/mympd -w $HOME/.config/mympd

```

**Reply by sandorex** [reply]
I am getting
```
Compiling myMPD
CANNOT LINK EXECUTABLE "cmake": cannot locate symbol "_ZNKSt6__ndk14__fs10filesystem4path10__filenameEv" referenced by "/data/data/com.termux/files/usr/bin/cmake"...
Aborted
```
Got it to build, i guess termux being funky

**Reply by sandorex** [reply]
Got it to work but had to disable SSL and use 8080 as port

**Comment by jcorporation** [comment]
Setting the environment variable LUA_DIR to the path of the lua installation should work. Without lua, the cool scripting features are disabled.

Anyway, this is the second attempt to run myMPD under termux, a howto in the documentation would be great. But I can not write it.

**Reply by sandorex** [reply]
I was originally planning on making some ci build the termux package and maybe even get it into the termux repos, we'll see how that works out though, no promises

**Comment by sandorex** [comment, marked as answer]
1. Update the packages, i had problems before i updated so do that first (it crashed few times for me but i just kept re running it)
```sh
pkg upgrade
```
2. Install dependencies
```sh
pkg install build-essential cmake perl pcre2 openssl libid3tag libflac lua54 git
```

#### Now to compile it
3. Clone the git repo (depth makes it download only the last commit making it much smaller and faster to download)
```sh
git clone https://github.com/jcorporation/myMPD.git --depth=1
```
4. Change directory to the git repo and run build
```sh
cd myMPD
export MYMPD_INSTALL_PREFIX=$PREFIX
export EXTRA_CMAKE_OPTIONS="-DLUA_MATH_LIBRARY=/system/lib64/libm.so" # substitute 'lib64' for 'lib' if you are on 32bit arch
./build.sh release
```

#### Initial config
5. Create the config directory for myMPD and run it to create the config files from the environment variables
```
mkdir -p $HOME/.config/mympd

export MYMPD_HTTP_PORT=8080 # you can use '80' if using root
export MYMPD_LOGLEVEL=6
export MPD_HOST=/data/data/com.termux/files/usr/var/run/mpd.socket # this is default for termux i think but check your /etc/mpd.conf value for bind_to_address
export MPD_PORT=35000 # choose whatever mpd you use in /etc/mpd.conf
export MYMPD_SSL=false # 'true' if using root
export MYMPD_SSL_PORT=443

$HOME/myMPD/release/mympd -w $HOME/.config/mympd # run it
```

#### Running it
After this run myMPD with just this, the rest is necessary for the first start only
```sh
$HOME/myMPD/release/mympd -w $HOME/.config/mympd
```

#### Running with root (if you want to use lower ports, or use SSL)
I made this little script which should run mympd as root with little trouble **(run it as the user!)**
```sh
#!/bin/bash
su root -c "$HOME/myMPD/release/mympd -w $HOME/.config/mympd -u $(whoami)"
```
You can also prepend `nohup` before `su` to make it start in background and not stop even after closing the terminal

Thanks a lot to @jcorporation for helping me get this compile properly without dirty hacks or patches 😄

**Reply by sandorex** [reply]
@jcorporation Do you think it's good enough for docs?

**Reply by jcorporation** [reply]
Looks great, only the libm hack is bad. Looking for termux in the cmake file is also no good idea. Could a externally set compile time option work?

**Reply by sandorex** [reply]
> Could a externally set compile time option work?

It probably would but do you know of any way to pass arguments to cmake through the build script? In the issue referenced they say `-DLUA_MATH_LIBRARY="/system/lib/libm.so"` fixed it, i don't recall testing this as environment variable `export LUA_MATH_LIBRARY=...` i will try that tonight hopefully but i did read FindLua cmake file and it does not look for this in the environment unfortunately

**Reply by jcorporation** [reply]
There was since commit 791e12ffacda6ced0a08602890f43fb7c0f62ea9 indeed no possibility. But with latest devel you could try following:
```
export EXTRA_CMAKE_OPTIONS="-DLUA_MATH_LIBRARY=/system/lib/libm.so"
```

**Reply by sandorex** [reply]
It works! Thank you very much for the help! 😄

**Reply by jcorporation** [reply]
I added your documentation to the official myMPD documentation for the next major release.

**Comment by roughnecks** [comment]
Hello.
posting here since it looks related.

While building I'm getting this:
```
Arch:
CMake Error at CMakeLists.txt:88 (find_library):
  Could not find MATH_LIB using the following names: m
```

Can you help?

**Reply by jcorporation** [reply]
You should open a new question with detailed description, environment and logs.
